### PR TITLE
Zenodo publishing

### PR DESCRIPTION
**Pull request by prioux**
This is operational, provided a task has been extended with the two methods that CBRAIN expects to create the Zenodo records. Two base CBRAIN tasks have this capability: Diagnostics and BashScriptor.

Soon to be added: automatic support for Boutiques tasks (in a separate PR?)

This will implement part of the requirements for #855 

Note: while testing it seems that Zenodo just refused all large files (greater than a few dozen megabytes).

### CODE REVIEW

**Comment by prioux** [comment]
Addendum:

The three commits I made today added zenodo support to most (all?) Boutiques tasks. Also, we now record and publish a lot more information about the runtime environment of the task.

To test with CIVET or ReconAll (traditional tasks) you'll need to deploy the PR for the cbrain-plugins-neuro : https://github.com/aces/cbrain-plugins-neuro/pull/146

File: BrainPortal/app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -771,6 +771,395 @@ def apply_operation(operation, taskids, do_in_spawn = false, options = {})
...
+    # with another DOI are just refered to as such.
+    related = @zenodo_userfiles
+      .select { |u| u.zenodo_doi.present? }
+      .map    { |u| ZenodoClient::RelatedIdentifier.new(:identifier => u.zenodo_doi, :relation => 'hasPart') }
+    if related.present?
+      @zenodo_metadata.related_identifiers ||= []
+      @zenodo_metadata.related_identifiers  += related
+    end
+
+    # Validate here; TODO
```
Comment (natacha-beck) [inline BrainPortal/app/controllers/tasks_controller.rb]: Should be done or remove no ?

File: BrainPortal/app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -771,6 +771,395 @@ def apply_operation(operation, taskids, do_in_spawn = false, options = {})
...
+  end
+
+  def init_zenodo_client(zsite) #:nodoc:
+    zsite  = zsite == 'main' ? 'main' : 'sandbox' # sanitize
+    token  = token_by_zsite(zsite)
+    server = zenodo_server_by_zsite(zsite)
+    ZenodoClient.configure do |config|
+      config.api_key['access_token'] = token
+      config.host                    = server
+      #config.debugging               = true # will print info on stdout; WARNING binary junk!
```
Comment (natacha-beck) [inline BrainPortal/app/controllers/tasks_controller.rb]: Could be removed?

File: BrainPortal/app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -771,6 +771,395 @@ def apply_operation(operation, taskids, do_in_spawn = false, options = {})
...
+    is_col = userfile.is_a?(FileCollection)
+
+    # Identify the content
+    userfile.sync_to_cache
+    content_path = userfile.cache_full_path
+    content_name = userfile.name
+    content_path = create_tmp_tar_for_filecollection(userfile) if is_col
+    content_name = userfile.name + ".tar.gz"                   if is_col
+
+    # Upload
```
Comment (natacha-beck) [inline BrainPortal/app/controllers/tasks_controller.rb]: Code duplication from 1105 to 1115 and from 1133 to 1143.

File: BrainPortal/app/helpers/zenodo_helper.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
+
+  ZenodoSandboxDOIPrefix = "10.5072/"
+  ZenodoMainDOIPrefix    = "10.5281/"
+
+  # Creates a pretty link to a DOI, usually for Zenodo
+  def link_to_zenodo_doi(doi)
+    link_to "<img src=\"https://zenodo.org/badge/DOI/#{doi}.svg\" alt=\"DOI\">".html_safe,
+            "https://doi.org/#{doi}", :target => '_blank'
+    #("<a href=\"https://doi.org/#{doi}\">" +
```
Comment (natacha-beck) [inline BrainPortal/app/helpers/zenodo_helper.rb]: Is it useful?

File: BrainPortal/app/controllers/tasks_controller.rb (hunk 1 above)
Comment (prioux) [reply BrainPortal/app/controllers/tasks_controller.rb]: Agreed. I'll add some validations, at least one.

File: BrainPortal/app/controllers/tasks_controller.rb (hunk 2 above)
Comment (prioux) [reply BrainPortal/app/controllers/tasks_controller.rb]: No, this is really useful, and someone who doesn't know how the ZenodoClient library works would probably be happy to know how to turn on debugging. I was using it all the time while developing, and I can imagine wanting to uncomment it when doing future work.

File: BrainPortal/app/controllers/tasks_controller.rb (hunk 3 above)
Comment (prioux) [reply BrainPortal/app/controllers/tasks_controller.rb]: Yes, but I'm not sure how to handle refactor it... will look into it.

File: BrainPortal/app/helpers/zenodo_helper.rb (hunk 4 above)
Comment (prioux) [reply BrainPortal/app/helpers/zenodo_helper.rb]: Nope, I'll remove that, thanks.

**Pull request merged**